### PR TITLE
Tag current deployed image into an image stream on deployment success

### PR DESCRIPTION
**Pull request by smarterclayton**
Add an "after" action set that can tag the deployed image into an image stream tag after the deployment succeeds. Allows an e2e style test to tag a :dev image into a :tested or :qe-ready tag (see transient PR #6930).

Use case: you run a deployment - if it succeeds, you tag the image into a "stable" branch for someone else to trigger on.

TODO:

* [x] tests for imagestreamtag update
* [x] e2e tests for tagging
* [x] decision on abort policy
* [x] structure of the after type - might be better as an array of after actions in order
* [x] better factoring of the code in deployer - should it be treated like a hook?  what abstraction do we need for future growth

@kargakis @ironcladlou @bparees

### CODE REVIEW

**Comment by smarterclayton** [comment]
@ironcladlou looking at the docs, why do we not support LifeCycleHandlerPolicy on Post?  I don't see any reason the code couldn't handle it today.

**Comment by smarterclayton** [comment]
First two commits are in their own PR, see the TODO list for open questions here.

**Comment by kargakis** [comment]
cc: @soltysh

File: api/swagger-spec/oapi-v1.json [hunk 1]
```diff
@@ -18197,6 +18197,9 @@
    },
    "v1.DeploymentStrategy": {
     "id": "v1.DeploymentStrategy",
+    "required": [
+     "after"
```
Comment (kargakis) [inline api/swagger-spec/oapi-v1.json]: Why is this required?

File: api/swagger-spec/oapi-v1.json [hunk 2]
```diff
@@ -18214,6 +18217,10 @@
       "$ref": "v1.RollingDeploymentStrategyParams",
       "description": "input to the Rolling deployment strategy"
      },
+     "after": {
+      "$ref": "v1.AfterDeploymentActions",
```
Comment (kargakis) [inline api/swagger-spec/oapi-v1.json]: How about `PostDeploymentActions`? More consistent with `Post` lifecycle hooks

File: pkg/cmd/infra/deployer/deployer.go [hunk 3]
```diff
@@ -196,5 +203,53 @@ func (d *Deployer) Deploy(namespace, deploymentName string) error {
...
+	after := config.Spec.Strategy.After
+	if after == nil {
+		return nil
+	}
+
+	var errs []error
+	if len(after.TagImages) > 0 {
+		glog.Infof("Tagging images for successful deployment")
+		for _, action := range after.TagImages {
+			image, err := deployapi.TemplateImageForContainer(&to.Spec.Template.Spec, deployapi.DeploymentConfigHasTrigger(config), action.ContainerName)
```
Comment (kargakis) [inline pkg/cmd/infra/deployer/deployer.go]: Seems like the replication controller is redundant since you can use the dc template

File: pkg/deploy/api/types.go [hunk 4]
```diff
@@ -28,12 +28,17 @@ const (
...
+
 	// CustomParams are the input to the Custom deployment strategy.
 	CustomParams *CustomDeploymentStrategyParams
 	// RecreateParams are the input to the Recreate deployment strategy.
 	RecreateParams *RecreateDeploymentStrategyParams
 	// RollingParams are the input to the Rolling deployment strategy.
 	RollingParams *RollingDeploymentStrategyParams
+
+	// After is an optional set of actions to take after a deployment has completed.
+	After *AfterDeploymentActions `json:"after" description:"actions to take after a deployment has completed"`
```
Comment (kargakis) [inline pkg/deploy/api/types.go]: remove the tags from here

File: pkg/deploy/api/types.go [hunk 5]
```diff
@@ -169,6 +174,19 @@ const (
 	DefaultRollingUpdatePeriodSeconds int64 = 1
 )
 
+type AfterDeploymentActions struct {
+	// TagImages instructs the deployer to tag the current image referenced under a container onto an image stream tag if the deployment succeeds.
+	TagImages []TagImageAction `json:"tagImages,omitempty" description:"a list of image stream tags that should be updated to the latest value on this deployment if the deployment succeeds"`
```
Comment (kargakis) [inline pkg/deploy/api/types.go]: remove the tags from here

File: pkg/deploy/api/v1/types.go [hunk 6]
```diff
@@ -28,12 +28,17 @@ const (
...
+
 	// CustomParams are the input to the Custom deployment strategy.
 	CustomParams *CustomDeploymentStrategyParams `json:"customParams,omitempty" description:"input to the Custom deployment strategy"`
 	// RecreateParams are the input to the Recreate deployment strategy.
 	RecreateParams *RecreateDeploymentStrategyParams `json:"recreateParams,omitempty" description:"input to the Recreate deployment strategy"`
 	// RollingParams are the input to the Rolling deployment strategy.
 	RollingParams *RollingDeploymentStrategyParams `json:"rollingParams,omitempty" description:"input to the Rolling deployment strategy"`
+
+	// After is an optional set of actions to take after a deployment has completed.
+	After *AfterDeploymentActions `json:"after" description:"actions to take after a deployment has completed"`
```
Comment (kargakis) [inline pkg/deploy/api/v1/types.go]: should be omitempty

File: pkg/cmd/infra/deployer/deployer.go [hunk 7]
```diff
@@ -196,5 +203,53 @@ func (d *Deployer) Deploy(namespace, deploymentName string) error {
 	} else {
 		glog.Infof("Deploying from %s to %s (replicas: %d)", deployutil.LabelForDeployment(from), deployutil.LabelForDeployment(to), desiredReplicas)
 	}
-	return strategy.Deploy(from, to, desiredReplicas)
+	if err := strategy.Deploy(from, to, desiredReplicas); err != nil {
+		return err
+	}
+	return d.after(config, to)
```
Comment (kargakis) [inline pkg/cmd/infra/deployer/deployer.go]: Why is this different from a post hook? Why should the deployment be marked as failed if our `AfterDeployment` actions fail?

File: api/swagger-spec/oapi-v1.json (hunk 1 above)
Comment (smarterclayton) [inline api/swagger-spec/oapi-v1.json]: Pointers should not be omit empty, this is a bug in swagger.

File: api/swagger-spec/oapi-v1.json (hunk 2 above)
Comment (smarterclayton) [inline api/swagger-spec/oapi-v1.json]: I was considering an array of hooks instead:

```
After []LifecycleHook
```

File: pkg/cmd/infra/deployer/deployer.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: Is the deployment template guaranteed to match the RC?  I don't think so.

File: api/swagger-spec/oapi-v1.json (hunk 1 above)
Comment (smarterclayton) [inline api/swagger-spec/oapi-v1.json]: A pointer is implicitly optional, we should not have to specify omit empty.

File: pkg/cmd/infra/deployer/deployer.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: I think post hooks should also be able to cause deployments to fail.  I don't think our original arguments still hold - basically, if I want to fail a deployment because of a post hook, I should be allowed to.  Same for After.  I do think after needs to be able to specify the abort policy, but I don't think the current Post behavior is right.

File: pkg/cmd/infra/deployer/deployer.go (hunk 3 above)
Comment (kargakis) [inline pkg/cmd/infra/deployer/deployer.go]: Well, the spec is a straight copy: https://github.com/openshift/origin/blob/733b7b76f455bf47cc617e4176afb29a14fdb601/pkg/deploy/util/util.go#L144 + https://github.com/openshift/origin/blob/733b7b76f455bf47cc617e4176afb29a14fdb601/pkg/deploy/util/util.go#L205

File: pkg/cmd/infra/deployer/deployer.go (hunk 7 above)
Comment (kargakis) [inline pkg/cmd/infra/deployer/deployer.go]: > basically, if I want to fail a deployment because of a post hook, I should be allowed to.

Agreed. This means hooks/after should hold an option for this.

File: pkg/cmd/infra/deployer/deployer.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: That's at the time the RC is created - not at the time it is processed by
the deployer.  If the RC is changed to *actually* use a different image,
that's what we want to use.  I.e. I want the thing that the majority of
pods used, as much as possible.

File: pkg/cmd/infra/deployer/deployer.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: Yeah, the array of after hooks with Lifecycle is probably best.  It means
lifecycle hooks can tag images - I think I'm ok with that.

Comment (ironcladlou) [inline pkg/cmd/infra/deployer/deployer.go]: Any reason not to make hooks arrays generally (i.e. pre, mid for recreate)?

Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: Other than it's a breaking change.  I'd say we can do it if someone asks,
but wouldn't do it until they ask.  I just want people to actually give us
feedback on hooks (good or bad) before we make more changes.  Mid is the
only feedback I've gotten, and they didn't even realize they needed it
themselves.

Comment (ironcladlou) [inline pkg/cmd/infra/deployer/deployer.go]: Sounds reasonable.

File: pkg/cmd/infra/deployer/deployer.go (hunk 3 above)
Comment (kargakis) [inline pkg/cmd/infra/deployer/deployer.go]: Where exactly the deployer is deviating the podspec of the latest rc from that of the dc? I would say that having a podspec in a rc different from the podspec in the dc is a bigger problem. I also think we are clear on direct manipulation of dc-owned rcs being a wrong thing.

Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: Is the deployment config guaranteed to have the latest image field set?  I
don't know that the annotation of the RC is guaranteed to have that - if
you're confident that the annotated DC on the RC is always correct, then
yes, we should use the value of the image coming from the annotated DC on
the RC.  But it shouldn't come from the DC itself.

File: pkg/deploy/api/types.go [hunk 8]
```diff
@@ -169,6 +174,19 @@ const (
...
+}
+
+// TagImageAction is a request to tag the image in a particular container onto an ImageStreamTag on a successful deployment.
+type TagImageAction struct {
+	// ContainerName is the name of a container in the deployment config whose image value will be used as the source of the tag
+	ContainerName string
+	// ImageStreamTagName is the name of an ImageStreamTag in the current namespace that will be updated if the deployment succeeds.
+	ImageStreamTagName string
```
Comment (bparees) [inline pkg/deploy/api/types.go]: why constraint it to this namespace?  why not allow the user to specify a name+namespace for the tag?

suppose I want to do all my TestDeployments in a particular namespace for quota reasons, but then tag the validated images to prod?

Comment (smarterclayton) [inline pkg/deploy/api/types.go]: The security check for tagging is harder.  However, I can kind of buy that
I might want to tag to other namespaces if I've granted my deployer the
permissions.

Comment (bparees) [inline pkg/deploy/api/types.go]: I guess it seems highly likely i'd want to run these sort of tests under a different quota than my prod quota, that's my primary scenario for wanting that capability.

File: pkg/cmd/infra/deployer/deployer.go (hunk 3 above)
Comment (kargakis) [inline pkg/cmd/infra/deployer/deployer.go]: I don't know of any other place where we touch rcs for dcs apart from the controllers. And there shouldn't be.
* [main dc controller](https://github.com/openshift/origin/blob/920ad9ab0697b9026152e5d358364533bd22eed1/pkg/deploy/controller/deploymentconfig/controller.go#L124)
* [configchange controller](https://github.com/openshift/origin/blob/920ad9ab0697b9026152e5d358364533bd22eed1/pkg/deploy/controller/configchange/controller.go#L79)

Both use `MakeDeployment` which does the encoding of the latest dc into the rc.

* the deployer pod and deployment controllers update with status annotations
* the image change controller doesn't even touch rcs

That being said, I perceive the DC as the state that the latest RC should follow and we should guarantee that (like what happens upstream). Right now we don't have any strong guarantees but I think it is the case.

Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: An admission controller that updated the RC to fully qualify pod references
(Or pods themselves) would be broken here.  I need to look at the state of
the deployment as it runs through.

File: api/swagger-spec/oapi-v1.json (hunk 2 above)
Comment (soltysh) [inline api/swagger-spec/oapi-v1.json]: Can't we just stick to Pre/Post names we already have at Strategy levels. The only difference is - those will be at Deployment level and not at strategy and explain to users how these relate.

Comment (smarterclayton) [inline api/swagger-spec/oapi-v1.json]: We can't change post to an array without breaking backwards compatibility.
PostPost is bad.  Having Post at two levels is confusing.  So we need to
have a name that is invoked after the strategy.  We could call it
PostStrategy, but I'm not sure that's easy to understand for users.

Comment (soltysh) [inline api/swagger-spec/oapi-v1.json]: That makes sense, than maybe BeforeAll/AfterAll that seems to have the connotation of being first/last.

Comment (kargakis) [inline api/swagger-spec/oapi-v1.json]: AtTheEndOfTheDay

File: pkg/deploy/api/v1/types.go (hunk 6 above)
Comment (ironcladlou) [inline pkg/deploy/api/v1/types.go]: What about:

```go
type DeploymentStrategy struct {
    // ..
    After []AfterDeploymentAction
    // ..
}

type AfterDeploymentAction struct {
    TagImage *TagImageAction
    // New actions can go here
}

type TagImageAction struct {
    // ...
}
```

Not only does it look much more obvious what actions you have when there is more than one action type defined, but you get ordered actions for free (if we want to support that on the processing side).

File: pkg/cmd/infra/deployer/deployer.go [hunk 9]
```diff
@@ -196,5 +203,53 @@ func (d *Deployer) Deploy(namespace, deploymentName string) error {
...
+
+// Handle processes the post-deployment phase
+func (a *AfterDeployment) Handle(config *deployapi.DeploymentConfig, to *kapi.ReplicationController) error {
+	after := config.Spec.Strategy.After
+	if after == nil {
+		return nil
+	}
+
+	var errs []error
+	if len(after.TagImages) > 0 {
```
Comment (ironcladlou) [inline pkg/cmd/infra/deployer/deployer.go]: Nit:

```go
if len(after.TagImages) == 0 {
    return nil
}
```

nestinglevel--

File: pkg/deploy/api/v1/types.go (hunk 6 above)
Comment (ironcladlou) [inline pkg/deploy/api/v1/types.go]: @smarterclayton not urgent, just want to make sure this suggestion isn't lost in the mix

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Ok, this has been simplified to simply add a TagImageHook array to LifecycleHook, as well as supporting code to enable tagging via imagestreamtag update (which now accepts a `tag` field to control the spec tag).

I've also made post hooks capable of failing the deployment - need to add a few more tests around them.

**Comment by smarterclayton** [comment]
TagImage now has a To object reference

**Comment by ironcladlou** [comment]
>Ok, this has been simplified to simply add a TagImageHook array to LifecycleHook, as well as supporting code to enable tagging via imagestreamtag update (which now accepts a tag field to control the spec tag).

It didn't even occur to me that this was a chance to introduce a new hook type... very nice!

>I've also made post hooks capable of failing the deployment - need to add a few more tests around them.

+1, shouldn't be problematic. We should maybe have done this sooner given our automatic rollback capabilities.

**Comment by ironcladlou** [comment]
Overall design now LGTM

**Comment by soltysh** [comment]
@smarterclayton rebase and you're good to go, LGTM from my side as well.

File: pkg/cmd/infra/deployer/deployer.go [hunk 10]
```diff
@@ -89,10 +91,10 @@ func NewDeployer(client kclient.Interface) *Deployer {
 		strategyFor: func(config *deployapi.DeploymentConfig) (strategy.DeploymentStrategy, error) {
 			switch config.Spec.Strategy.Type {
 			case deployapi.DeploymentStrategyTypeRecreate:
-				return recreate.NewRecreateDeploymentStrategy(client, kapi.Codecs.UniversalDecoder()), nil
+				return recreate.NewRecreateDeploymentStrategy(client, oclient, kapi.Codecs.UniversalDecoder()), nil
```
Comment (deads2k) [inline pkg/cmd/infra/deployer/deployer.go]: @liggitt convinced me.  Casting the `oClient` down into the type we're expecting makes this easier to follow.

Comment (deads2k) [inline pkg/cmd/infra/deployer/deployer.go]: > @liggitt convinced me. Casting the oClient down into the type we're expecting makes this easier to follow.

`client.ImageStreamTagsNamespacer(oclient)`

File: pkg/deploy/api/types.go [hunk 11]
```diff
@@ -77,17 +79,20 @@ type RecreateDeploymentStrategyParams struct {
...
+// LifecycleHook defines a specific deployment lifecycle action. Only one type of action may be specified at any time.
 type LifecycleHook struct {
 	// FailurePolicy specifies what action to take if the hook fails.
 	FailurePolicy LifecycleHookFailurePolicy
+
 	// ExecNewPod specifies the options for a lifecycle hook backed by a pod.
 	ExecNewPod *ExecNewPodHook
+
+	// TagImages instructs the deployer to tag the current image referenced under a container onto an image stream tag if the deployment succeeds.
+	TagImages []TagImageHook
```
Comment (deads2k) [inline pkg/deploy/api/types.go]: Shouldn't this only be allowed on `Post` hooks?  That suggests a different type to me.

File: pkg/deploy/api/types.go [hunk 12]
```diff
@@ -118,6 +123,14 @@ type ExecNewPodHook struct {
 	Volumes []string
 }
 
+// TagImageHook is a request to tag the image in a particular container onto an ImageStreamTag.
+type TagImageHook struct {
+	// ContainerName is the name of a container in the deployment config whose image value will be used as the source of the tag
+	ContainerName string
```
Comment (deads2k) [inline pkg/deploy/api/types.go]: This is defaultable when we only have one container.  I don't see that happening and it should be mentioned in the godoc.

**Comment by deads2k** [comment]
@kargakis This change will require an update to `oc status` to perform a cycle check.

File: pkg/deploy/api/validation/validation.go [hunk 13]
```diff
@@ -123,33 +127,56 @@ func validateCustomParams(params *deployapi.CustomDeploymentStrategyParams, fldP
...
+		errs = append(errs, field.Invalid(fldPath, "<hook>", "only one of 'execNewPod' of 'tagImages' may be specified"))
+	case hook.ExecNewPod != nil:
 		errs = append(errs, validateExecNewPod(hook.ExecNewPod, fldPath.Child("execNewPod"))...)
+	case len(hook.TagImages) > 0:
+		for i, image := range hook.TagImages {
+			if len(image.ContainerName) == 0 {
+				errs = append(errs, field.Required(fldPath.Child("tagImages").Index(i).Child("containerName"), "a containerName is required"))
+			} else {
+				if _, err := deployapi.TemplateImageForContainer(pod, deployapi.IgnoreTriggers, image.ContainerName); err != nil {
+					errs = append(errs, field.Invalid(fldPath.Child("tagImages").Index(i).Child("containerName"), image.ContainerName, err.Error()))
```
Comment (deads2k) [inline pkg/deploy/api/validation/validation.go]: You should be able to give me back a list of valid values.  `NotSupported` is the wrong connotation, but the right shape of the error.

File: pkg/deploy/api/validation/validation.go [hunk 14]
```diff
@@ -123,33 +127,56 @@ func validateCustomParams(params *deployapi.CustomDeploymentStrategyParams, fldP
...
+				}
+			}
+			if image.To.Kind != "DockerImage" {
```
Comment (deads2k) [inline pkg/deploy/api/validation/validation.go]: Why `DockerImage` and not `ImageStreamTag`?

Comment (deads2k) [inline pkg/deploy/api/validation/validation.go]: `ImageStreamTag` gives parity with the `DeploymentTriggerImageChangeParams` (forced to only istag in validation) and allows a very basic cycle check to see if tagging would feed this DC.

File: pkg/deploy/api/validation/validation_test.go [hunk 15]
```diff
@@ -371,6 +371,146 @@ func TestValidateDeploymentConfigMissingFields(t *testing.T) {
...
 		},
+		"missing spec.strategy.recreateParams.mid.execNewPod": {
+			api.DeploymentConfig{
+				ObjectMeta: kapi.ObjectMeta{Name: "foo", Namespace: "bar"},
+				Spec: api.DeploymentConfigSpec{
+					Replicas: 1,
+					Strategy: api.DeploymentStrategy{
+						Type: api.DeploymentStrategyTypeRecreate,
+						RecreateParams: &api.RecreateDeploymentStrategyParams{
+							Mid: &api.LifecycleHook{
```
Comment (deads2k) [inline pkg/deploy/api/validation/validation_test.go]: They're like little green Christmas trees in my diff. :)

File: pkg/deploy/strategy/recreate/recreate.go [hunk 16]
```diff
@@ -50,7 +52,7 @@ const AcceptorInterval = 1 * time.Second
 
 // NewRecreateDeploymentStrategy makes a RecreateDeploymentStrategy backed by
 // a real HookExecutor and client.
-func NewRecreateDeploymentStrategy(client kclient.Interface, decoder runtime.Decoder) *RecreateDeploymentStrategy {
+func NewRecreateDeploymentStrategy(client kclient.Interface, tags client.ImageStreamTagsNamespacer, decoder runtime.Decoder) *RecreateDeploymentStrategy {
```
Comment (deads2k) [inline pkg/deploy/strategy/recreate/recreate.go]: `tagClient`?  I would generally think of `tags` as actual tag objects.

File: pkg/deploy/api/types.go [hunk 17]
```diff
@@ -77,17 +79,20 @@ type RecreateDeploymentStrategyParams struct {
 	// pod is created. All LifecycleHookFailurePolicy values are supported.
 	Mid *LifecycleHook
 	// Post is a lifecycle hook which is executed after the strategy has
-	// finished all deployment logic. The LifecycleHookFailurePolicyAbort policy
```
Comment (deads2k) [inline pkg/deploy/api/types.go]: This change is really strange.  Specifying `Abort` here doesn't seem to what I'd expect.  The action has already happened, right?  The new RC is up and running.  I may have marked it as failed, but its actually there answering my requests, isn't it?

I didn't think that we had compensation logic to handle an automatic rollback in the case of a deployment marked failure because it didn't matter before, you couldn't abort after the action was done.

File: pkg/deploy/strategy/support/lifecycle.go [hunk 18]
```diff
@@ -83,6 +91,46 @@ func (e *HookExecutor) Execute(hook *deployapi.LifecycleHook, deployment *kapi.R
...
+			continue
+		}
+		value := image.Image
+		namespace := action.To.Namespace
+		if len(namespace) == 0 {
+			namespace = deployment.Namespace
+		}
+		if _, err := e.tags.ImageStreamTags(namespace).Update(&imageapi.ImageStreamTag{
+			ObjectMeta: kapi.ObjectMeta{
+				Name:      action.To.Name,
```
Comment (deads2k) [inline pkg/deploy/strategy/support/lifecycle.go]: Yeah, I'm pretty convinced that `hook.TagImages.To.Kind` should be `ImageStreamTag`.  That's what we're updating here.

File: pkg/image/api/types.go [hunk 19]
```diff
@@ -194,6 +194,19 @@ type ImageStreamTag struct {
 	unversioned.TypeMeta
 	kapi.ObjectMeta
 
+	// Tag is the spec tag associated with this image stream tag, and it may be null
+	// if only pushes have occured to this image stream.
+	Tag *TagReference
```
Comment (deads2k) [inline pkg/image/api/types.go]: Are these mutually exclusive to `Image`?

File: pkg/image/api/validation/validation.go [hunk 20]
```diff
@@ -159,18 +170,33 @@ func ValidateImageStreamMapping(mapping *api.ImageStreamMapping) field.ErrorList
 
 // ValidateImageStreamTag is essentially a no-op.  We don't allow direct creation of istags
 func ValidateImageStreamTag(ist *api.ImageStreamTag) field.ErrorList {
-	return validation.ValidateObjectMeta(&ist.ObjectMeta, true, oapi.MinimalNameRequirements, field.NewPath("metadata"))
+	result := validation.ValidateObjectMeta(&ist.ObjectMeta, true, oapi.MinimalNameRequirements, field.NewPath("metadata"))
+	if ist.Tag != nil {
+		result = append(result, ValidateImageStreamTagReference(*ist.Tag, field.NewPath("tag"))...)
+		if ist.Tag.Annotations != nil && !kapi.Semantic.DeepEqual(ist.Tag.Annotations, ist.ObjectMeta.Annotations) {
+			result = append(result, field.Invalid(field.NewPath("tag", "annotations"), "<map>", "tag annotations must not be provided or must be equal to the object meta annotations"))
```
Comment (deads2k) [inline pkg/image/api/validation/validation.go]: This is a really unusual restriction.  Wouldn't this mean that a future `oc annotate` call would forever fail on these objects?

File: pkg/deploy/api/types.go (hunk 11 above)
Comment (kargakis) [inline pkg/deploy/api/types.go]: Agreed

Comment (kargakis) [inline pkg/deploy/api/types.go]: If we will not use a different type, we should at least disallow them during pre and mid validation. But I still feel this is a different type.

File: pkg/image/api/types.go [hunk 21]
```diff
@@ -194,6 +194,19 @@ type ImageStreamTag struct {
...
+
+	// Generation is the current generation of the tagged image - if tag is provided
+	// and this value is not equal to the tag generation, a user has requested an
+	// import that has not completed, or Conditions will be filled out indicating any
+	// error.
+	Generation int64
```
Comment (kargakis) [inline pkg/image/api/types.go]: Why can't we use objectmeta's generation?

File: pkg/image/api/validation/validation.go [hunk 22]
```diff
@@ -159,18 +170,33 @@ func ValidateImageStreamMapping(mapping *api.ImageStreamMapping) field.ErrorList
 
 // ValidateImageStreamTag is essentially a no-op.  We don't allow direct creation of istags
```
Comment (kargakis) [inline pkg/image/api/validation/validation.go]: update this comment

File: pkg/image/registry/imagestreamtag/rest.go [hunk 23]
```diff
@@ -212,12 +239,25 @@ func newISTag(tag string, imageStream *api.ImageStream, image *api.Image) (*api.
 			Annotations:       map[string]string{},
 			ResourceVersion:   imageStream.ResourceVersion,
 		},
+		Generation: event.Generation,
+		Conditions: imageStream.Status.Tags[tag].Conditions,
 	}
 
-	// if the imageStream has Spec.Tags[tag].Annotations[k] = v, copy it to the image's annotations
-	// and add them to the istag's annotations
 	if imageStream.Spec.Tags != nil {
```
Comment (kargakis) [inline pkg/image/registry/imagestreamtag/rest.go]: I don't think we need this check since we are not mutating

File: pkg/image/registry/imagestreamtag/rest.go [hunk 24]
```diff
@@ -190,18 +212,23 @@ func (r *REST) Delete(ctx kapi.Context, id string) (runtime.Object, error) {
...
 	if event == nil || len(event.Image) == 0 {
-		return nil, kapierrors.NewNotFound(api.Resource("imagestreamtag"), api.JoinImageStreamTag(imageStream.Name, tag))
+		return nil, kapierrors.NewNotFound(api.Resource("imagestreamtags"), api.JoinImageStreamTag(imageStream.Name, tag))
 	}
 
 	return r.imageRegistry.GetImage(ctx, event.Image)
 }
 
-func newISTag(tag string, imageStream *api.ImageStream, image *api.Image) (*api.ImageStreamTag, error) {
+func newISTag(tag string, imageStream *api.ImageStream, image *api.Image, allowEmptyEvent bool) (*api.ImageStreamTag, error) {
```
Comment (kargakis) [inline pkg/image/registry/imagestreamtag/rest.go]: godoc

**Comment by smarterclayton** [comment]
Flake 

```
ERROR: gave up waiting for https://127.0.0.1:38443/healthz/ready

!!! Error in /data/src/github.com/openshift/origin/hack/../hack/util.sh:370
	'return 1' exited with status 1
Call stack:
```

merging [merge]

**Comment by smarterclayton** [comment]
I see you snuck in and added comments... clever.

File: pkg/cmd/infra/deployer/deployer.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/cmd/infra/deployer/deployer.go]: He hasn't convinced me

File: pkg/deploy/api/types.go (hunk 11 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: No reason it has to be - I don't see anything special about post hooks.
For instance, as pre I could tag it into a test to do something different.

File: pkg/deploy/api/types.go (hunk 12 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: Is defaultable, don't see it being necessary to default it.  I'd rather
people be explicit.  However, I will validate it for people

File: pkg/deploy/api/validation/validation.go (hunk 13 above)
Comment (smarterclayton) [inline pkg/deploy/api/validation/validation.go]: We don't do that in all places, and names can be very long, so I'm not sure
we need to start introducing it here.

File: pkg/deploy/api/validation/validation.go (hunk 14 above)
Comment (smarterclayton) [inline pkg/deploy/api/validation/validation.go]: Bug.

File: pkg/deploy/strategy/recreate/recreate.go (hunk 16 above)
Comment (smarterclayton) [inline pkg/deploy/strategy/recreate/recreate.go]: Yeah

File: pkg/deploy/api/types.go (hunk 17 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: In a failure case, the scaler reconciles the old back to the new - a failed
deployment gets rolled back by scaling.

File: pkg/image/api/types.go (hunk 19 above)
Comment (smarterclayton) [inline pkg/image/api/types.go]: No, this is the spec tag

File: pkg/image/api/validation/validation.go (hunk 20 above)
Comment (smarterclayton) [inline pkg/image/api/validation/validation.go]: No, annotations are reconciled with prepareupdate.  Basically, you can't
change annotations via tags because "back compat"

File: pkg/deploy/api/types.go (hunk 11 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: No, and no.

File: pkg/image/api/types.go (hunk 21 above)
Comment (smarterclayton) [inline pkg/image/api/types.go]: Spec tags have their own validation.

Comment (smarterclayton) [inline pkg/image/api/types.go]: Spec tags have their own *generation*

File: pkg/deploy/api/types.go (hunk 11 above)
Comment (deads2k) [inline pkg/deploy/api/types.go]: > No reason it has to be - I don't see anything special about post hooks.
> For instance, as pre I could tag it into a test to do something different.

Then your doc is wrong.  It will tag it regardless of the deployment succeeds or not.

File: pkg/deploy/api/types.go (hunk 12 above)
Comment (deads2k) [inline pkg/deploy/api/types.go]: > Is defaultable, don't see it being necessary to default it.  I'd rather
> people be explicit.  However, I will validate it for people

The entire API and CLI are based around the idea of specifying as little information as possible.  Why is this case special?  The "normal" flow is a DC with a single container.

File: pkg/deploy/api/validation/validation.go (hunk 14 above)
Comment (deads2k) [inline pkg/deploy/api/validation/validation.go]: > Bug.

This isn't clear to me.  You're saying that `DeploymentTriggerImageChangeParams`, which we went to great effort to make trigger only on `ImageStreamTag` is wrong and that you want to explicitly require that someone say they are updating a `DockerImage` when you explicitly use the value to update an `ImageStreamTag` is the right thing to do?  Can you provide some more detail?

File: pkg/deploy/api/types.go (hunk 12 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: Ok

File: pkg/deploy/api/validation/validation.go (hunk 14 above)
Comment (smarterclayton) [inline pkg/deploy/api/validation/validation.go]: TagImages has a bug.

**Comment by smarterclayton** [comment]
[merge]
